File: dec1/main.py

import sys

def get_location_id_lists(path: str) -> tuple[list[int], list[int]]: 
    left_numbers = []
    right_numbers = []

    with open(path) as f:
        for line in f:
            numbers = [int(x) for x in line.split()]
            left_numbers.append(numbers[0])
            right_numbers.append(numbers[1])

    sorted_left = sorted(left_numbers)
    sorted_right = sorted(right_numbers)

    return sorted_left, sorted_right

def get_distance_score(sorted_left: list[int], sorted_right: list[int]) -> list[int]:
    distance_score = 0

    for pair in zip(sorted_left, sorted_right):
        distance_score += abs(pair[0] - pair[1])

    return distance_score

def get_similarity_score(sorted_left: list[int], sorted_right: list[int]) -> int:
    frequency_dict = create_frequency_dict(sorted_right)
    similarity_score = 0
    for number in sorted_left:
        if number in frequency_dict:
            similarity_score += number * frequency_dict[number]
            
    return similarity_score
    
def create_frequency_dict(numbers: list[int]) -> dict[int, int]:
    frequency_dict = {}
    for number in numbers:
        if number in frequency_dict:
            frequency_dict[number] += 1
        else:
            frequency_dict[number] = 1
    return frequency_dict

def main():
    sorted_left, sorted_right = get_location_id_lists(sys.argv[1])
    distance_score = get_distance_score(sorted_left, sorted_right)
    similarity_score = get_similarity_score(sorted_left, sorted_right)

    print("Distance   score: ", distance_score)
    print("Similarity score: ", similarity_score)

if __name__ == "__main__":
    main()

File: dec1/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"slices"
	"strconv"
	"strings"
)

func main() {
	// Read file and prepare number lists
	filePath := os.Args[1]
	lines := read_file(filePath)
	left, right := build_lists_from_lines(lines)

	// Run calculations
	distance_score := get_distance_score(left, right)
	similarity_score := get_similarity_score(left, right)
	fmt.Println("Distance score  : ",distance_score)
	fmt.Println("Similarity score: ",similarity_score)
}

func read_file(path string) []string {
	readFile, err := os.Open(path);

	if err != nil {
		fmt.Println(err)
	}

	fileScanner := bufio.NewScanner(readFile)
    fileScanner.Split(bufio.ScanLines)
    var fileLines []string

	for fileScanner.Scan() {
		fileLines = append(fileLines, fileScanner.Text())
	}

	readFile.Close()

	return fileLines
}

func build_lists_from_lines(lines []string) ([]int, []int) {
	var left_numbers []int
	var right_numbers []int

	for i := 0; i < len(lines); i++ {
		var numbers = strings.Fields(lines[i])
		left_num, err_left := strconv.Atoi(numbers[0])
		right_num, err_right := strconv.Atoi(numbers[1])

		if err_left != nil || err_right != nil {
			panic("Invalid input file")
		}

		
		left_numbers = append(left_numbers, left_num)
		right_numbers = append(right_numbers, right_num)


		slices.Sort(left_numbers)
		slices.Sort(right_numbers)
		
	}
	return left_numbers, right_numbers
}

func get_distance_score(sorted_left []int, sorted_right []int) int {
	var distance_score int
	for i := 0; i < len(sorted_left); i++ {
		distance_score += abs(sorted_left[i] - sorted_right[i])
	}
	return distance_score
}

func get_similarity_score(sorted_left []int, sorted_right []int) int {
	var similarity_score int
	var frequency_map = get_frequency_map(sorted_right)

	for i := 0; i < len(sorted_left); i++ {
		if _, ok := frequency_map[sorted_left[i]]; ok {
			similarity_score += sorted_left[i] * frequency_map[sorted_left[i]]
		}
	}
	
	return similarity_score
}

func get_frequency_map(numbers []int) map[int]int {
	frequency_map := make(map[int]int)
	for i := 0; i < len(numbers); i++ {
		if _, ok := frequency_map[numbers[i]]; ok {
			frequency_map[numbers[i]]++
		} else {
			frequency_map[numbers[i]] = 1
		}
	}
	return frequency_map
}

func abs(a int) int {
    if a >= 0 {
        return a
    }
    return -a
}